server/src/transport/ws.rs

```diff
@@ -564,10 +565,13 @@ async fn graceful_shutdown(
 			//
 			// The receiver is not cancel-safe such that it's used in a stream to enforce that.
 			let disconnect_stream = futures_util::stream::unfold((receiver, data), |(mut receiver, mut data)| async {
-				if let Err(SokettoError::Closed) = receiver.receive(&mut data).await {
-					None
-				} else {
-					Some(((), (receiver, data)))
+				match receiver.receive(&mut data).await {
+					Ok(Incoming::Closed(_)) | Err(SokettoError::Closed) => None,
+					Ok(Incoming::Data(_) | Incoming::Pong(_)) => Some(((), (receiver, data))),
+					Err(e) => {
```

Comment:
Just to check; do you no longer need to return `None` when `Err(SokettoError::Closed)` is a thing now?

Reply:
We will keep the stream alive until either `SokettoError::Closed` or `Incoming::Closed` is returned back because some implementations may not do the proper `WebSocket close handshake` message such as the soketto client https://github.com/paritytech/jsonrpsee/issues/871. Thus, we need to check both

However, soketto will eventually return SokettError::Closed after this https://github.com/paritytech/soketto/blob/master/src/connection.rs#L386-L388 so this extra check in this PR shouldn't matter in "practice" but we need to call receive one more time instead then also I like us to be explicit here which makes it easier to understand.

FWIW, it may also a bunch of I/O errors that could terminate the connection as well but those are super-tricky to detect so we are relying on the `close stuff`...

Reply:
Oohh I actually missed that you were still paying attention to `SokettoError::Closed` anyway, so yup all makes sense to me now :)